sql/stats: generate statistics forecasts

michae2 opened the pull request:
**sql/stats: use nil eval.Context as CompareContext when forecasting**

When forecasting table statistics, we don't need a full *eval.Context.
We can simply use a nil *eval.Context as a tree.CompareContext. This
means we don't have to plumb an eval.Context into the stats cache.

Assists: #79872

Release note: None

**sql/stats: generate statistics forecasts in the stats cache**

As of this commit, we now try to generate statistics forecasts for every
column of every table. This happens whenever statistics are loaded into
or refreshed in the stats cache. We use only the forecasts that fit the
historical collected statistics very well, meaning we have high
confidence in their accuracy.

Fixes: #79872

Release note (performance improvement): Enable table statistics
forecasts, which predict future statistics based on historical collected
statistics. Forecasts help the optimizer produce better plans for
queries that read data modified after the latest statistics collection.
We use only the forecasts that fit the historical collected statistics
very well, meaning we have high confidence in their accuracy. Forecasts
can be viewed using `SHOW STATISTICS FOR TABLE ... WITH FORECAST`.

**sql: show forecasted stats time in EXPLAIN**

When using statistics forecasts, add the forecast time (which could be
in the future) to EXPLAIN output. This both indicates that forecasts are
in use, and gives us an idea of how up-to-date / ahead they are.

Assists: #79872

Release note: None

**sql/opt: add tests for statistics forecasts**

Add a few simple testcases for usage of statistics forecasts by the
optimizer.

Assists: #79872

Release note: None

---

Release justification: Enable feature before we get too far into
stability period.

michae2 committed:
sql/stats: use nil *eval.Context as CompareContext when forecasting

When forecasting table statistics, we don't need a full *eval.Context.
We can simply use a nil *eval.Context as a tree.CompareContext. This
means we don't have to plumb an eval.Context into the stats cache.

Assists: #79872

Release note: None

michae2 committed:
sql/stats: generate statistics forecasts in the stats cache

As of this commit, we now try to generate statistics forecasts for every
column of every table. This happens whenever statistics are loaded into
or refreshed in the stats cache. We use only the forecasts that fit the
historical collected statistics very well, meaning we have high
confidence in their accuracy.

Fixes: #79872

Release note (performance improvement): Enable table statistics
forecasts, which predict future statistics based on historical collected
statistics. Forecasts help the optimizer produce better plans for
queries that read data modified after the latest statistics collection.
We use only the forecasts that fit the historical collected statistics
very well, meaning we have high confidence in their accuracy. Forecasts
can be viewed using `SHOW STATISTICS FOR TABLE ... WITH FORECAST`.

michae2 committed:
sql: show forecasted stats time in EXPLAIN

When using statistics forecasts, add the forecast time (which could be
in the future) to EXPLAIN output. This both indicates that forecasts are
in use, and gives us an idea of how up-to-date / ahead they are.

Assists: #79872

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/86078)
<!-- Reviewable:end -->

michae2 committed:
sql/opt: add tests for statistics forecasts

Add a few simple testcases for usage of statistics forecasts by the
optimizer.

Assists: #79872

Release note: None

michae2 commented:
Opening this to get some CI testing.

michae2 commented:
Ok, this is RFAL.

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 5 of 6 files at r1, 12 of 12 files at r5, 2 of 2 files at r6, 7 of 7 files at r7, 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86078#-:-N9YThn_CjTmk6leGXvI:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @michae2, and @yuzefovich)

___
*[`pkg/sql/opt/exec/explain/emit.go` line 495 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86078#-N9YSVAUDPlBt2BcRTPQ:-N9YSVAUDPlBt2BcRTPR:bqk3ow4) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7c577cd056e7733be8f1d91b43038da278558d4/pkg/sql/opt/exec/explain/emit.go#L495)):*
> ```Go
> 							}
> 							duration = fmt.Sprintf(
> 								"; stats collected %s ago", humanizeutil.LongDuration(timeSinceStats),
> ```

I feel like it might still be useful to report how long ago stats were collected, even if we're using a forecast -- what do you think? Would it be too much info?
___
*[`pkg/sql/stats/forecast.go` line 259 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/86078#-N9YQCj_9HI-q6rx4-8W:-N9YQCj_9HI-q6rx4-8X:b4hj96f) ([raw file](https://github.com/cockroachdb/cockroach/blob/135c9762356ba51cd49ef048ccaed1549bb19015/pkg/sql/stats/forecast.go#L259)):*
> ```Go
> 
> 		hist, err := predictHistogram(
> 			ctx, evalCtx, observed, forecastAt, minRequiredFit, nonNullRowCount,
> ```

Why don't you also remove it from this function?
___
*[`pkg/sql/stats/quantile_test.go` line 49 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/86078#-N9YQJETDpAIxaRqsyqq:-N9YQJETDpAIxaRqsyqr:b-tmsav3) ([raw file](https://github.com/cockroachdb/cockroach/blob/135c9762356ba51cd49ef048ccaed1549bb19015/pkg/sql/stats/quantile_test.go#L49)):*
> ```Go
> 			for i := 0; i < 5; i++ {
> 				t.Run(fmt.Sprintf("%v/%v", colType.Name(), i), func(t *testing.T) {
> 					hist, rowCount := randHist(evalCtx, colType, rng)
> ```

Can you remove `evalCtx` from `randHist`?
___
*[`pkg/sql/stats/quantile_test.go` line 571 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/86078#-N9YQrDM7Jy1zsfYYIW3:-N9YQrDM7Jy1zsfYYIW4:b68vca5) ([raw file](https://github.com/cockroachdb/cockroach/blob/135c9762356ba51cd49ef048ccaed1549bb19015/pkg/sql/stats/quantile_test.go#L571)):*
> ```Go
> 	for i, tc := range testCases {
> 		t.Run(strconv.Itoa(i), func(t *testing.T) {
> 			hist, err := tc.qfun.toHistogram(evalCtx, types.Float, tc.rows)
> ```

Can you remove `evalCtx` from `toHistogram`?
___
*[`pkg/sql/stats/quantile_test.go` line 869 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/86078#-N9YQksY1J_JVVXfPUB8:-N9YQksY1J_JVVXfPUB9:bf3rrh) ([raw file](https://github.com/cockroachdb/cockroach/blob/135c9762356ba51cd49ef048ccaed1549bb19015/pkg/sql/stats/quantile_test.go#L869)):*
> ```Go
> 				return
> 			}
> 			cmp, err := res.CompareError(evalCtx, tc.dat)
> ```

Can you just pass `nil` here?
___
*[`pkg/sql/stats/quantile_test.go` line 1136 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/86078#-N9YQpEg6B_f9e4zUDMc:-N9YQpEg6B_f9e4zUDMd:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/135c9762356ba51cd49ef048ccaed1549bb19015/pkg/sql/stats/quantile_test.go#L1136)):*
> ```Go
> 				return
> 			}
> 			cmp, err := d.CompareError(evalCtx, tc.dat)
> ```

ditto


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 5 of 6 files at r1, 12 of 12 files at r5, 2 of 2 files at r6, 7 of 7 files at r7, 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86078#-:-N9YUvCX7iffTCFtv4CC:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @mgartner and @michae2)

___
*[`pkg/sql/opt/exec/execbuilder/testdata/forecast` line 168 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/86078#-N9YUYKK-VWCyYYcFfpi:-N9YUYKK-VWCyYYcFfpj:b-upk586) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4b95991d47ee53f6e88c15662124beb9d810b83/pkg/sql/opt/exec/execbuilder/testdata/forecast#L168)):*
> ```
> ]'
> 
> query TTTIIII
> ```

Can this test be flaky because the stats cache hasn't been loaded?


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86078)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner, @michae2, @rytaft, and @yuzefovich)

___
*[`pkg/sql/opt/exec/execbuilder/testdata/forecast` line 168 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/86078#-N9YUYKK-VWCyYYcFfpi:-N9Zv-xp9AHUSD1xmB1B:b1zzhc2) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4b95991d47ee53f6e88c15662124beb9d810b83/pkg/sql/opt/exec/execbuilder/testdata/forecast#L168)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Can this test be flaky because the stats cache hasn't been loaded?
</blockquote></details>

Querying a table will force the stats cache to load the stats for that table, so I don't think that type of flake is possible, IIU.
___
*[`pkg/sql/opt/exec/explain/emit.go` line 495 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/86078#-N9YSVAUDPlBt2BcRTPQ:-N9ZvqB5AiWT5ROuHRR4:b-s2d6ob) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7c577cd056e7733be8f1d91b43038da278558d4/pkg/sql/opt/exec/explain/emit.go#L495)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I feel like it might still be useful to report how long ago stats were collected, even if we're using a forecast -- what do you think? Would it be too much info?
</blockquote></details>

I like it! Done.
___
*[`pkg/sql/stats/forecast.go` line 259 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/86078#-N9YQCj_9HI-q6rx4-8W:-N9Zvu2m8ojI1C2O6N7i:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/135c9762356ba51cd49ef048ccaed1549bb19015/pkg/sql/stats/forecast.go#L259)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Why don't you also remove it from this function?
</blockquote></details>

Done.
___
*[`pkg/sql/stats/quantile_test.go` line 49 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/86078#-N9YQJETDpAIxaRqsyqq:-N9ZvvVgEvXU7ptBi9GT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/135c9762356ba51cd49ef048ccaed1549bb19015/pkg/sql/stats/quantile_test.go#L49)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Can you remove `evalCtx` from `randHist`?
</blockquote></details>

Done.
___
*[`pkg/sql/stats/quantile_test.go` line 571 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/86078#-N9YQrDM7Jy1zsfYYIW3:-N9Zvwd4EJewmQ0n6Kqk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/135c9762356ba51cd49ef048ccaed1549bb19015/pkg/sql/stats/quantile_test.go#L571)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Can you remove `evalCtx` from `toHistogram`?
</blockquote></details>

Done.
___
*[`pkg/sql/stats/quantile_test.go` line 869 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/86078#-N9YQksY1J_JVVXfPUB8:-N9ZvxYa2kj0s1uIJW4k:b-apwzv9) ([raw file](https://github.com/cockroachdb/cockroach/blob/135c9762356ba51cd49ef048ccaed1549bb19015/pkg/sql/stats/quantile_test.go#L869)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Can you just pass `nil` here?
</blockquote></details>

It needs to be a typed `nil`, so I don't think I can.


<!-- Sent from Reviewable.io -->

michae2 commented:
TFTRs!

bors r=rytaft,yuzefovich
